### ISSUE TITLE
vms_apis: live migration of vm api doesn't work 

### ISSUE THREAD

**Issue by ahmedelsayed-93** [zero-os/0-orchestrator#215]
### API
```POST nodes/{{nodeid}}/vms/{{vmid}}/migrate```

### request & response
![image](https://cloud.githubusercontent.com/assets/24386820/26031904/64258c62-3884-11e7-9660-7bda0062aff8.png)

### ays log
```  
- error during execution of step 1 in run ed5fb429cc7bba0ee9663ed99608944c
Error of job: vm!vm (migrate):
*TRACEBACK*********************************************************************************
Traceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/9b66e201ec9cd6eac3cec055faa491a0.py", line 245, in migrate
    raise j.exceptions.Input("migrate action expect to have the destination node in the argument")
JSExceptions.Input: ERROR: migrate action expect to have the destination node in the argument ((type:input.error)


******************************************************************************************

[Sun14 07:02] - Job.py              :54  :j.jobcontroller.job.4ab47e8d19974ae7d28cc2e412253f32 - ERROR    - Traceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/9b66e201ec9cd6eac3cec055faa491a0.py", line 99, in processChange
    update_data(job, args)
  File "/tmp/actions/vm/9b66e201ec9cd6eac3cec055faa491a0.py", line 341, in update_data
    service.model.data.memory = args['memory']
KeyError: 'memory'
```

### SV
js ```441b344```
resourcepool ```da19c73 ```

**Comment by Dinaamagdy** [comment]
###  It doesn't give error as shown in api request : 
![selection_356](https://cloud.githubusercontent.com/assets/24386802/26404020/c0e7282a-4090-11e7-8413-ceada89fa397.png)
###  but vm doesn't migrate : 
![selection_357](https://cloud.githubusercontent.com/assets/24386802/26404330/d1cc70a4-4091-11e7-8e15-aa903021721a.png)

### And  it gives error in ays and makes it stop : 
```
[Wed24 12:50] - 455b6fc291e7bd2e7afb2bf4b098ef99.py:100 :j.jobcontroller.job.9e4c8bc31eb88a94512c57680a088db9 - INFO     - killing process /bin/nbdserver
did not receive an Exceptio object for python exception, this is serious bug.
exceptionObject was:
[invalid type, expecting <class 'int'>] at -> //pid
did not receive an Exceptio object for python exception, this is serious bug.
exceptionObject was:
1
2017-05-24 12:50:10 - (sanic)[INFO]: Stopping worker [29040]
2017-05-24 12:50:10 - (sanic)[ERROR]: Experienced exception while trying to serve
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/sanic/server.py", line 450, in serve
    loop.run_forever()
  File "uvloop/loop.pyx", line 1100, in uvloop.loop.Loop.run_forever (uvloop/loop.c:24056)
  File "uvloop/loop.pyx", line 393, in uvloop.loop.Loop._run (uvloop/loop.c:11845)
  File "uvloop/loop.pyx", line 311, in uvloop.loop.Loop._on_idle (uvloop/loop.c:10730)
  File "uvloop/cbhandles.pyx", line 49, in uvloop.loop.Handle._run (uvloop/loop.c:48380)
  File "/opt/jumpscale8/lib/JumpScale/baselib/jobcontroller/Job.py", line 50, in _execute_cb
    eco = j.errorconditionhandler.processPythonExceptionObject(ex)
  File "/opt/jumpscale8/lib/JumpScale/core/errorhandling/ErrorConditionHandler.py", line 126, in processPythonExceptionObject
    exceptionObject=exceptionObject, tb=tb)
  File "/opt/jumpscale8/lib/JumpScale/core/errorhandling/ErrorConditionHandler.py", line 161, in parsePythonExceptionObject
    sys.exit(1)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/sanic/app.py", line 559, in run
    serve(**server_settings)
  File "/usr/local/lib/python3.5/dist-packages/sanic/server.py", line 459, in serve
    loop.run_until_complete(http_server.wait_closed())
  File "uvloop/loop.pyx", line 1198, in uvloop.loop.Loop.run_until_complete (uvloop/loop.c:25480)
  File "uvloop/loop.pyx", line 1191, in uvloop.loop.Loop.run_until_complete (uvloop/loop.c:25324)
  File "uvloop/loop.pyx", line 1100, in uvloop.loop.Loop.run_forever (uvloop/loop.c:24056)
  File "uvloop/loop.pyx", line 393, in uvloop.loop.Loop._run (uvloop/loop.c:11845)
  File "uvloop/loop.pyx", line 311, in uvloop.loop.Loop._on_idle (uvloop/loop.c:10730)
  File "uvloop/cbhandles.pyx", line 49, in uvloop.loop.Handle._run (uvloop/loop.c:48380)
  File "uvloop/future.pyx", line 448, in uvloop.loop.BaseTask._wakeup (uvloop/loop.c:114340)
  File "uvloop/future.pyx", line 441, in uvloop.loop.BaseTask._fast_wakeup (uvloop/loop.c:114043)
  File "uvloop/future.pyx", line 383, in uvloop.loop.BaseTask._fast_step (uvloop/loop.c:113654)
  File "uvloop/future.pyx", line 372, in uvloop.loop.BaseTask._fast_step (uvloop/loop.c:112669)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 392, in wait_for
    return fut.result()
  File "uvloop/future.pyx", line 146, in uvloop.loop.BaseFuture.result (uvloop/loop.c:109361)
  File "uvloop/future.pyx", line 101, in uvloop.loop.BaseFuture._result_impl (uvloop/loop.c:108900)
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/a2f7d8168f52904b8293f918709ead62.py", line 271, in migrate
    j.tools.async.wrappers.sync(nbdserver.executeAction('stop'))
  File "/opt/jumpscale8/lib/JumpScale/tools/async/Wrappers.py", line 11, in sync
    return loop.run_until_complete(coro)
  File "/usr/lib/python3.5/asyncio/base_events.py", line 375, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1312, in _run_once
    handle._run()
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/opt/jumpscale8/lib/JumpScale/baselib/jobcontroller/Job.py", line 50, in _execute_cb
    eco = j.errorconditionhandler.processPythonExceptionObject(ex)
  File "/opt/jumpscale8/lib/JumpScale/core/errorhandling/ErrorConditionHandler.py", line 126, in processPythonExceptionObject
    exceptionObject=exceptionObject, tb=tb)
  File "/opt/jumpscale8/lib/JumpScale/core/errorhandling/ErrorConditionHandler.py", line 161, in parsePythonExceptionObject
    sys.exit(1)
SystemExit: 1
```

**Comment by Dinaamagdy** [comment]
INFO:root:Cleaning up...
Traceback (most recent call last):
  File "bd-performance.py", line 349, in <module>
    test_fio_nbd()
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 87, in augment_usage_errors
    yield
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "bd-performance.py", line 59, in test_fio_nbd
    cleanUp(api, nodeIDs, deployInfo)
  File "bd-performance.py", line 136, in cleanUp
    waitProcess(api, disconnectDiskCommand, jobId, nodeID, testContainer)
  File "bd-performance.py", line 308, in waitProcess
    raise RuntimeError("Command %s failed to execute successfully. %s" % (command, res["stderr"]))
RuntimeError: Command {'pwd': '', 'args': ['-d', '/dev/nbd0'], 'name': '/bin/nbd-client'} failed to execute successfully. Error: Ioctl failed: Invalid argument
Exiting.

**Comment by grimpy** [comment]
@Dinaamagdy this comment does not belong to this ticket I guess?

**Comment by abdulrahmantkhalifa** [comment]
- So it was decided that we are going to use ssh for the live migration , instead of tcp as it was more secure. 
 - The 0-orchestrator code reaches the command of the pyclient kvm.migrate and returns and error.
 - The error basicly is becuase the netcat lib we use  does not support unix sockets so a new issue in 0-initramfs was created https://github.com/zero-os/0-initramfs/issues/54, this issue is in alpha 6 milestone so i dont know how this will be finished in alpha 5 since it is blocked on it

**Comment by abdulrahmantkhalifa** [comment]
migrate works for me , the problem is  that the change parent does not remove the original service from the old parent https://github.com/Jumpscale/ays9/issues/111

**Comment by Dinaamagdy** [comment]
- Status of vm after migration is migrating not running 
- vm exist in  in  two hosts  
- can't create vm again in nodes and give this error : 
```
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/AtYourServiceDependencies.py", line 63, in addEdges
    addEdges(edge, action, all_nodes, nodes)
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/AtYourServiceDependencies.py", line 58, in addEdges
    for producers in node.service.producers.values():
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 398, in producers
    producers[role].append(self.aysrepo.serviceGet(key=prod.key))
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/AtYourServiceRepo.py", line 338, in serviceGet
    return self.db.services.services[key]
RecursionError: maximum recursion depth exceeded
```
## Software Vsersion 
7faeed68dc37804ae2229c52ceacfd04807c0f79

**Comment by abdulrahmantkhalifa** [comment]
cannot reproduce "Status of vm after migration is migrating not running"

**Comment by Dinaamagdy** [comment]
Now , it return 204 in Api request but ays throw this error in ays and node doesn't migrate  
```
INFO:network:
[Wed26 14:04] - Job.py              :57  :j.core.jobcontroller.job.7603e5f0743eee43fa564220e809030f - ERROR    - OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/3214b086c3a4f4c28ab2642f1c6390a3.py", line 439, in processChange
    update_data(job, args)
  File "/tmp/actions/vm/3214b086c3a4f4c28ab2642f1c6390a3.py", line 93, in update_data
    j.tools.async.wrappers.sync(job.execute())
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 352, in execute
    self.save()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 308, in save
    self.service.saveAll()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 329, in saveAll
    self.saveToFS()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 319, in saveToFS
    j.sal.fs.writeFile(path3, self.model.dataJSON)
  File "/opt/code/github/jumpscale/core9/JumpScale9/fs/SystemFS.py", line 1288, in writeFile
    fp.close()
OSError: [Errno 9] Bad file descriptor
```
and after that I  can't create vdisk any more with this error:
```
   producers[role].append(self.aysrepo.serviceGet(key=prod.key))
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/AtYourServiceRepo.py", line 338, in serviceGet
    return self.db.services.services[key]
RecursionError: maximum recursion depth exceede
```
## Steps to reproduce 
- create storagecluster for 2 nodes
- create vdisk for this st
- create vm in node 1
- migrate to second node
## Vm request 
{
  "id": "vm",
  "memory": 1024,
  "cpu": 1,
	"disks": [
    {
      "vdiskid": "vdisk",
      "maxIOps": 2000   
    }
  ],
  "userCloudInit": null,
  "systemCloudInit": null
} 
## Vdisk request
```
{
	"id": "vdisk",
	"size": 15,
	"blocksize": 4096,
	"type": "boot",
	"templatevdisk": "ardb://hub.gig.tech:16379/template:ubuntu-1604",
	"readOnly": false,
	"storagecluster": "st"
}
```
## Software Version 
cce7cc1e1f957e0eb691b863502fa6c3f4620d52

**Comment by abdulrahmantkhalifa** [comment]
the problem is that the change parent does not remove the original service from the old parent Jumpscale/ays9#111 this is Blocking

**Comment by Dinaamagdy** [comment]
vm after migration  exist in two nodes 
## Software Version 
28055ea5f6706956585ec2e3c69c07c7ab13abf8

**Comment by Dinaamagdy** [comment]
## Error
```
INFO:network:
[Wed02 14:49] - Job.py              :56  :j.core.jobcontroller.job.633dd26b0f9442ce23b217c252de2c44 - ERROR    - Traceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/4176b98f5fa2cc3c421638bb9ba61dee.py", line 473, in processChange
    update_data(job, args)
  File "/tmp/actions/vm/4176b98f5fa2cc3c421638bb9ba61dee.py", line 26, in update_data
    j.tools.async.wrappers.sync(job.execute())
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 354, in execute
    self.save()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 310, in save
    self.service.saveAll()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 329, in saveAll
    self.saveToFS()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 319, in saveToFS
    j.sal.fs.writeFile(path3, self.model.dataJSON)
  File "/opt/code/github/jumpscale/core9/JumpScale9/fs/SystemFS.py", line 1279, in writeFile
    fp = open(filename, "wb")
OSError: [Errno 9] Bad file descriptor
```
and machine still exist in 2 nodes 
## Software Version
 3784e265fb38bef07fd616b8d34aae2d7fc01b43

**Comment by Dinaamagdy** [comment]
it gives this error now
```
[Thu03 11:16] - Job.py              :56  :j.core.jobcontroller.job.47a4096bc07372bc57c1990e02c82290 - ERROR    - Tr
aceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/actions/vm/024a906089a002c3ec6f4d3ba880d920.py", line 113, in processChange
    update_data(job, args)
  File "/tmp/actions/vm/024a906089a002c3ec6f4d3ba880d920.py", line 700, in update_data
    j.tools.async.wrappers.sync(job.execute())
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 354, in execute
    self.save()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/jobcontroller/Job.py", line 310, in save
    self.service.saveAll()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 329, in saveAll
    self.saveToFS()
  File "/opt/code/github/jumpscale/ays9/JumpScale9AYS/ays/lib/Service.py", line 316, in saveToFS
    j.sal.fs.writeFile(path2, self.model.dictJson, append=False)
  File "/opt/code/github/jumpscale/core9/JumpScale9/fs/SystemFS.py", line 1284, in writeFile
    fp.write(bytes(contents, 'UTF-8'))
OSError: [Errno 9] Bad file descriptor
```
and vm status in new node " migrating"
## Software Version
ac3e7ec3b5f2e8f09ba32b6e4889053103830a66

**Comment by Dinaamagdy** [comment]
## Software Version
dd33720692713295f3a6ada7e33333bb1c8f3f4a
